pkg/otelcollector/config/provider.go

```diff
@@ -103,26 +89,54 @@ func (p *Provider) setConfig(config *Config) {
 // The hook should treat the given config as temporary.
 // The hook will also be executed immediately, to ensure that current config
 // was passed through all the added hooks.
+//
+// WARNING: This is supposed to be called only during initialization.
 func (p *Provider) AddMutatingHook(hook func(*Config)) {
-	p.lock.Lock()
-	defer p.lock.Unlock()
-
 	if p.config == nil {
 		log.Warn().Msg("OtelConfigProvider.AddHook: already shut down")
 		return
 	}
 
 	p.hooks = append(p.hooks, hook)
 
-	// Now the config provided to collector is outdated, run the newly-added hook.
-	hook(p.config)
-	p.setConfig(p.config)
+	p.UpdateConfig(p.config)
+}
+
+// GetConfig returns the current config.
+func (p *Provider) GetConfig() *Config {
+	p.configLock.RLock()
+	defer p.configLock.RUnlock()
+	return p.config
+}
+
+// SetConfig sets the new config, replacing the old one.
+func (p *Provider) SetConfig(config *Config) {
+	p.configLock.Lock()
+	defer p.configLock.Unlock()
+	p.config = config
+}
+
+// SetConfigIfNotNil sets the new config only if it nil.
+func (p *Provider) SetConfigIfNotNil(config *Config) {
+	p.configLock.Lock()
+	defer p.configLock.Unlock()
+	if p.config == nil {
+		log.Warn().Msg("OtelConfigProvider: tried to update config after Shutdown")
+		return
+	}
+	p.config = config
+}
+
+// GetWatchFunc returns the current watch function.
+func (p *Provider) GetWatchFunc() confmap.WatcherFunc {
+	p.watchFuncLock.Lock()
+	defer p.watchFuncLock.Unlock()
```

Comment:
what's the rationale for locking only in getters instead of top-level functions? This makes it easier to break invariants